### PR TITLE
✨ Handle NameEntered action

### LINKED ISSUE #171: Add the NameEntered action
## Description
We need to add the action that changes the name of the editable project as the user types

## Dependencies
Depends on both #169 and #170 

## Definition of done
- [ ] The NameEntered action is added to the list of Project actions
- [ ] The action is handled by the reducers
- [ ] The behavior is tested
- [ ] The action is emited by the UI
- [ ] The documentation is updated to reflect the changes done in this PR

### PR DESCRIPTION
**Pull request by pdrj**
### Summary
This handles NameEntered action.

### Why is it done like this and not some other way?
It's pretty straightforward.

### Relationships

Closes #171 

### Pull Request checklist

- [x] The code does not hinder the app's performance
- [x] The code does not introduce regressions
- [x] The code in this PR is tested thoroughly enough that changing the behavior to an incorrect one breaks tests

### Pointers
Trigger the fragment in `MainActivity`: `ProjectDialogFragment().show(supportFragmentManager, "test")`

Add some logging in the project reducer. Validate you receive the action when typing in the project name field.

### Review template

- [ ] The code is thoroughly tested in well written and resilient manner
- [ ] The code is performant and leads to an optimal user experience
- [ ] The code is testable and does not cut corners for the sake of delivering
- [ ] The code considers older platforms well and will work in older devices just like it will in new ones

- [ ] The code considers light and dark theme
- [ ] The code considers RTL and does not mention left/right
- [ ] The code considers i18n
  - [ ] There are no hardcoded strings in the UI
  - [ ] All strings are included in every language we support OR there's an issue requesting new translations
- [ ] There are no hard-coded numbers referring to dimensions or font sizes
- [ ] The code considers UI elements that are similar and extracts their similarities into styles to avoid duplication

## Merge permissions
Anyone

### CODE REVIEW

File: timer/src/main/java/com/toggl/timer/project/domain/ProjectReducer.kt [hunk 1]
```diff
@@ -18,6 +18,10 @@ class ProjectReducer @Inject constructor(
         action: ProjectAction
     ): List<Effect<ProjectAction>> =
         when (action) {
-            is ProjectAction.NameEntered -> noEffect()
+            is ProjectAction.NameEntered -> {
+                val editableProject = state.value.editableProject!!
```
Comment (heytherewill) [inline timer/src/main/java/com/toggl/timer/project/domain/ProjectReducer.kt:22]: I'd rather see and `state.value.editableProject ?: throw InvalidStateOperation("Project can never be null")`, which is easier to track if we ever get to it than a null reference exception

File: timer/src/main/java/com/toggl/timer/project/domain/ProjectReducer.kt [hunk 2]
```diff
@@ -18,6 +18,10 @@ class ProjectReducer @Inject constructor(
...
+                state.value = state.value.copy(editableProject = editableProject.copy(name = action.name))
```
Comment (heytherewill) [inline timer/src/main/java/com/toggl/timer/project/domain/ProjectReducer.kt:23]: Nested copying can be avoided by using arrow. See how in the start edit reducer

File: timer/src/main/java/com/toggl/timer/project/domain/ProjectReducer.kt [hunk 3]
```diff
@@ -18,6 +18,12 @@ class ProjectReducer @Inject constructor(
         action: ProjectAction
     ): List<Effect<ProjectAction>> =
         when (action) {
-            is ProjectAction.NameEntered -> noEffect()
+            is ProjectAction.NameEntered -> {
+                state.value.editableProject ?: throw IllegalStateException("editableProject must not be null")
```
Comment (daividssilverio) [inline timer/src/main/java/com/toggl/timer/project/domain/ProjectReducer.kt:22]: Oh, gotta use the new notation. This should look like:
```
is ProjectAction.NameEntered -> state.mutateWithoutEffects {
                editableProject ?: throw IllegalStateException("editableProject must not be null")
                ProjectState.editableProject.modify(this) {
                    it.copy(name = action.name)
                }
            }
```

Comment (daividssilverio) [reply timer/src/main/java/com/toggl/timer/project/domain/ProjectReducer.kt:22]: Ok... I forgot to set an `editableProject` to the state before testing... it started crashing of course.

**Pull request merged**